satpy/readers/modis_l2.py

```diff
@@ -104,112 +109,116 @@ def read_geo_resolution(metadata):
-
-        # No byte manipulation required
+            dataset = self._extract_and_mask_category_dataset(dataset_id, dataset_info, dataset_name_in_file)
         else:
-            dataset = self.load_dataset(dataset_name_in_file)
+            # No byte manipulation required
+            dataset = self.load_dataset(dataset_name_in_file, dataset_info.pop("category", False))
 
         self._add_satpy_metadata(dataset_id, dataset)
         return dataset
 
+    def _extract_and_mask_category_dataset(self, dataset_id, dataset_info, var_name):
+        # what dimension is per-byte
+        byte_dimension = None if self.is_imapp_mask_byte1 else dataset_info['byte_dimension']
+        dataset = self._select_hdf_dataset(var_name, byte_dimension)
+        # category products always have factor=1/offset=0 so don't apply them
+        # also remove them so they don't screw up future satpy processing
+        dataset.attrs.pop('scale_factor', None)
+        dataset.attrs.pop('add_offset', None)
+        # Don't do this byte work if we are using the IMAPP mask_byte1 file
+        if self.is_imapp_mask_byte1:
+            return dataset
+
+        dataset = _extract_byte_mask(dataset,
+                                     dataset_info['byte'],
+                                     dataset_info['bit_start'],
+                                     dataset_info['bit_count'])
+        dataset = self._mask_with_quality_assurance_if_needed(dataset, dataset_info, dataset_id)
+        return dataset
+
+    def _mask_with_quality_assurance_if_needed(self, dataset, dataset_info, dataset_id):
+        if not dataset_info.get('quality_assurance', False):
+            return dataset
+
+        # Get quality assurance dataset recursively
+        quality_assurance_dataset_id = dataset_id.from_dict(
+            dict(name='quality_assurance', resolution=1000)
+        )
+        quality_assurance_dataset_info = {
+            'name': 'quality_assurance',
+            'resolution': 1000,
+            'byte_dimension': 2,
+            'byte': 0,
+            'bit_start': 0,
+            'bit_count': 1,
+            'file_key': 'Quality_Assurance'
+        }
+        quality_assurance = self.get_dataset(
+            quality_assurance_dataset_id, quality_assurance_dataset_info
+        )
+        # Duplicate quality assurance dataset to create relevant filter
+        duplication_factor = [int(dataset_dim / quality_assurance_dim)
+                              for dataset_dim, quality_assurance_dim
+                              in zip(dataset.shape, quality_assurance.shape)]
+        quality_assurance = np.tile(quality_assurance, duplication_factor)
+        # Replace unassured data by NaN value
+        dataset[np.where(quality_assurance == 0)] = dataset.attrs["_FillValue"]
+        return dataset
+
 
-def bits_strip(bit_start, bit_count, value):
+def _extract_byte_mask(dataset, byte_information, bit_start, bit_count):
```

Comment:
There is also a bit mask extraction class in the olci reader, maybe we could factorize these two in the future?

Reply:
Yes we could probably. I didn't actually implement any of this I just extracted it from the method in the file handler. I'm sure there are some small differences, but I have no feelings about factorizing in the future.